src/canisters/frontend/ic-certified-assets/assets.did

```diff
@@ -77,6 +77,12 @@ type StreamingStrategy = variant {
   };
 };
 
+type SetAssetPropertiesArguments = record {
+  key: Key;
+  max_age: opt nat;
+  headers: opt vec HeaderField;
+};
```

Comment:
Let's discuss what the semantics of this call should be, but afterwards, we should document those semantics here.

Reply:
do you think something like this:
```
type SetAssetPropertiesArguments = record {
  key: Key;
  max_age: opt opt nat;
  headers: opt opt vec HeaderField;
};
```
would work better? `opt opt` would allow to:
- `null` - indicates the value should not change
- `opt null` - set the value to `None`
- `opt opt 5` - set the value to `Some(5)`

Reply:
This is what I would suggest.  What do you think?
```
pub enum AssetProperty {
    MaxAge(Option<u64>),
    Headers(Option<HashMap<String, String>>),
}
pub struct SetAssetPropertiesArguments {
    pub key: Key,
    pub properties: Vec<AssetProperty>,
}
```

Reply:
I'm not sure -- the thing I don't like about this solution is it allows the user to send
```
SetAssetPropertiesArguments {
  key: "/index.html",
  properties: [MaxAge(Some(1)), MaxAge(Some(2)), MaxAge(Some(3))]
}
```
which is resolvable, but doesn't feel smooth to me

Reply:
Sure.  In this case, the answer to that is "don't do that."

Reply:
what is in your view the advantage of using `Vec<Enum>` for this?

Reply:
For example, the compiler can ensure that the code handles all of the cases.  Not so with n Options.
But I think it's mostly aesthetics.  Imagine if in the future if there were 8 properties.
I think overall I don't like any of the solutions yet.  What would you think about `SetAssetMaxAge` and `SetAssetHeaders` ?

Reply:
> For example, the compiler can ensure that the code handles all of the cases. Not so with n Options.
> ...
> Imagine if in the future if there were 8 properties.

Personally, I'd much rather have code poorly optimized by the compiler, than ask the user to spend x8 more cycles to update the properties of an asset. Of course, we can try guessing how often the user will try to update ALL properties (vs just one), but we won't know for sure until we have telemetry.

To be honest, I'm also not too concerned with aesthetics here - this is a utility method and not the core part of the frontend canister public interface, and notably, in 99% of the cases it will be utilized by `ic_asset::sync via dfx deploy`, and not via manual `dfx canister call` or similar

I'm also not in love with any of the solutions, but I like `opt opt` better than 
- `Vec<Enum>`, because vec enum has ambiguity (allows for dups, when dups are present order matters)
- having one `set` method for each asset property, because that burns more cycles 

Sadly I don't have better ideas, but there are other tasks requiring my focus, so unless you have some other ideas, I'd like to move forward with currently proposed implementation

Reply:
> code poorly optimized by the compiler

Nothing above has anything to do with compiler optimization.  Rather, that `match` can ensure that all possible cases are covered.

That said, just go ahead with the `opt opt` approach.

Reply:
> Rather, that match can ensure that all possible cases are covered.

sorry, I don't think I follow, can you explain? I don't get how would that be different from:

https://github.com/dfinity/sdk/blob/e2e186dfb4c8da7b33443c0703da863fbfda9c69/src/canisters/frontend/ic-certified-assets/src/state_machine.rs#L577-L582

Reply:
Happy to.  Referring to the fact that the compiler will report an error if one skips an enum type in a `match`, whereas it won't notice a missing `if let Some(...)` statement for one of many `opt opt` fields.  Of course tests should catch this also, but that was what I was referring to by `the compiler can ensure that the code handles all of the cases.`

```
match (prop) {
    MaxAge(max_age) => asset.max_age = max_age,
    Headers(headers) => asset.headers = headers,
    Redirect(url) => asset.redirect = redirect, // compiler error to leave this line out
}
```
Anyway, there aren't any right answers here, and `opt opt` may even be better in terms of backwards compatibility in the interface.